tools/pylib/boututils/datafile.py

```diff
@@ -571,49 +597,100 @@ def keys(self):
     def dimensions(self, varname):
         """Array of dimension names"""
         var = self.handle[varname]
-        vartype = str(var.attrs['type'],encoding='utf-8')
+        vartype = str(var.attrs['type'], encoding='utf-8')
         if vartype == 'Field3D_t':
-            return ('t','x','y','z')
+            return ('t', 'x', 'y', 'z')
         elif vartype == 'Field2D_t':
-            return ('t','x','y')
+            return ('t', 'x', 'y')
         elif vartype == 'scalar_t':
             return ('t')
         elif vartype == 'Field3D':
-            return ('x','y','z')
+            return ('x', 'y', 'z')
         elif vartype == 'Field2D':
-            return ('x','y')
+            return ('x', 'y')
         elif vartype == 'scalar':
             return ()
         else:
             raise ValueError("Variable type not recognized")
 
+    def vartype_from_array(self, data):
+        try:
+            ndim = len(data.shape)
+        except AttributeError:
+            ndim = 0
+        if ndim == 4:
+            return 'Field3D_t'
+        elif ndim == 3:
+            # not ideal, 3d field might be time-evolving 2d field,
+            # 'Field2D_t', but can't think of a good way to distinguish
```

Comment:
Please could you put this in the docstring? It feels like it could be a bit of a "gotcha"

Reply:
By itself probably no easy way to distinguish -- could possibly add an optional that represents the time array so could then check if the first dim has the same length as that, but that might start getting a bit ugly.

Reply:
I think a longer-term solution should be to make more use of metadata in the dmp files: whether a variable is time-evolving, units etc.

Reply:
This is actually a method on `Datafile`, so we could just look up `t_array` in the file and compare with that.

+1 for attributes. Not 100% sure how best to do it in for pylib though. One way might be to wrap `numpy.ndarray` so we can add our own attributes.

Reply:
It's possible to define our own `dtype` which includes names so it _might_ be possible to define the relevant dtypes for the different types of data, but this might not be the nicest.